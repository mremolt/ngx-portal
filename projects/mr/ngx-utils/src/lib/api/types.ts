import { HttpErrorResponse, HttpRequest } from '@angular/common/http';
import { Action } from '@ngrx/store';

export interface Constructor<T> {
  new (...args: any[]): T;
}

export interface ApiTokenSelector {
  select(state: any): string;
}

export interface ApiTokenSelectorConstructor {
  new (): ApiTokenSelector;
}

export interface ApiStartAction extends Action {
  readonly payload?: any;
}

export interface ApiSuccessAction<T> extends Action {
  readonly payload: T;
}

export interface ApiErrorAction extends Action {
  readonly payload: HttpErrorResponse;
}

export type ApiActions<T> = Readonly<{
  start: Constructor<ApiStartAction>;
  success: Constructor<ApiSuccessAction<T>>;
  error: Constructor<ApiErrorAction>;
  complete?: Constructor<Action>;
}>;

export type ApiRequestConfig<T> = Readonly<{
  actions: ApiActions<T>;
  request: HttpRequest<T>;
}>;
