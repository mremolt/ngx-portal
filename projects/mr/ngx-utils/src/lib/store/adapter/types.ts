import { DeepReadonly } from 'utility-types';

export type NormalizedEntities<Entity extends object> = DeepReadonly<{
  collection: { [key: string]: Entity };
  [key: string]: { [key: string]: object };
}>;

export type CollectionState<Entity extends object> = DeepReadonly<{
  entities: NormalizedEntities<Entity>;
  result: string[];
  selected: string | null;
  loading: boolean;
  loaded: boolean;
  dirty: boolean;
  updatedAt: Date | null;
}>;
