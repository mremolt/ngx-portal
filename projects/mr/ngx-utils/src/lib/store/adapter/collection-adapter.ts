import { schema } from 'normalizr';

import { remove, removeMany, updateEmbedded, upsertFactory, upsertManyFactory } from './reducers';
import { CollectionState } from './types';

export function buildCollectionState<Entity extends object>(): CollectionState<Entity> {
  return {
    entities: { collection: {} },
    result: [],
    selected: null,
    loading: false,
    loaded: false,
    dirty: false,
    updatedAt: null,
    error: null,
  };
}

export function createCollectionAdapter<Entity extends object>(entitySchema: schema.Entity) {
  type State = CollectionState<Entity>;

  const initialState = buildCollectionState<Entity>();

  const reducers = {
    upsert: upsertFactory(entitySchema),
    upsertMany: upsertManyFactory(entitySchema),
    updateEmbedded,
    remove,
    removeMany,

    reset(_state: State) {
      return initialState;
    },
  };

  return { initialState, reducers };
}
