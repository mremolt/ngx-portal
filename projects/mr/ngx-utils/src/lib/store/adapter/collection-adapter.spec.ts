import { createCollectionAdapter } from './collection-adapter';
import { User, userSchema, users } from './test.fixtures';
import { CollectionState } from './types';

const { user1, user2, user3 } = users;

describe('createCollectionAdapter', () => {
  describe('collectionAdapter', () => {
    const subject = createCollectionAdapter<User>(userSchema);

    it('creates the adapter', () => {
      expect(subject).toBeTruthy();
    });

    describe('initialState', () => {
      it('returns the correct structured state', () => {
        expect(subject.initialState).toMatchSnapshot();
      });
    });

    describe('reducers', () => {
      describe('upsert', () => {
        it('adds the entity to an empty state', () => {
          const newState = subject.reducers.upsert(subject.initialState, user1);
          expect(newState).toMatchSnapshot();
        });
      });

      describe('upsertMany', () => {
        it('adds the entities to an empty state', () => {
          const newState = subject.reducers.upsertMany(subject.initialState, [user1, user2]);
          expect(newState).toMatchSnapshot();
        });
      });

      describe('reset', () => {
        it('returns the initialState', () => {
          let newState = subject.reducers.upsertMany(subject.initialState, [user1, user2]);
          expect(newState).not.toBe(subject.initialState);

          newState = subject.reducers.reset(newState);
          expect(newState).toBe(subject.initialState);
        });
      });

      describe('remove', () => {
        it('removes an entity from entities and id from result', () => {
          const currentState = subject.reducers.upsertMany(subject.initialState, [
            user1,
            user2,
            user3,
          ]);
          const newState = subject.reducers.remove(currentState, user2.id);

          expect(newState).not.toEqual(currentState);
          expect(newState).toMatchSnapshot();
        });
      });

      describe('removeMany', () => {
        it('removes selected entities from entities and ids from result', () => {
          const currentState = subject.reducers.upsertMany(subject.initialState, [
            user1,
            user2,
            user3,
          ]);
          const newState = subject.reducers.removeMany(currentState, [user2.id, user1.id]);

          expect(newState).not.toEqual(currentState);
          expect(newState).toMatchSnapshot();
        });
      });

      describe('updateEmbedded', () => {
        let currentState: CollectionState<User>;

        beforeEach(() => {
          currentState = subject.reducers.upsertMany(subject.initialState, [user1, user2]);
        });

        it('updates the embedded entity', () => {
          const newState = subject.reducers.updateEmbedded(currentState, 'hobbies', {
            id: '12',
            name: 'drinking coffee',
          });
          expect(newState).toMatchSnapshot();
        });
      });
    });
  });
});
