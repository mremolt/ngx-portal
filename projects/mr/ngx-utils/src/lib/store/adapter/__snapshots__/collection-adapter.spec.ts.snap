// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createCollectionAdapter collectionAdapter initialState returns the correct structured state 1`] = `
Object {
  "dirty": false,
  "entities": Object {
    "collection": Object {},
  },
  "loaded": false,
  "loading": false,
  "result": Array [],
  "selected": null,
  "updatedAt": null,
}
`;

exports[`createCollectionAdapter collectionAdapter reducers remove removes an entity from entities and id from result 1`] = `
Object {
  "dirty": true,
  "entities": Object {
    "collection": Object {
      "42": Object {
        "firstname": "Arthur",
        "hobbies": Array [
          "12",
          "13",
        ],
        "id": "42",
        "lastname": "Dent",
        "planet": "43",
      },
      "46": Object {
        "firstname": "Ford",
        "hobbies": Array [
          "13",
        ],
        "id": "46",
        "lastname": "Perfect",
        "planet": "44",
      },
    },
    "hobbies": Object {
      "12": Object {
        "id": "12",
        "name": "Drinking tea",
      },
      "13": Object {
        "id": "13",
        "name": "Carrying towels",
      },
    },
    "planets": Object {
      "43": Object {
        "id": "43",
        "name": "Earth",
      },
      "44": Object {
        "id": "44",
        "name": "Beteigeuze",
      },
    },
  },
  "loaded": false,
  "loading": false,
  "result": Array [
    "42",
    "46",
  ],
  "selected": null,
  "updatedAt": null,
}
`;

exports[`createCollectionAdapter collectionAdapter reducers removeMany removes selected entities from entities and ids from result 1`] = `
Object {
  "dirty": true,
  "entities": Object {
    "collection": Object {
      "46": Object {
        "firstname": "Ford",
        "hobbies": Array [
          "13",
        ],
        "id": "46",
        "lastname": "Perfect",
        "planet": "44",
      },
    },
    "hobbies": Object {
      "12": Object {
        "id": "12",
        "name": "Drinking tea",
      },
      "13": Object {
        "id": "13",
        "name": "Carrying towels",
      },
    },
    "planets": Object {
      "43": Object {
        "id": "43",
        "name": "Earth",
      },
      "44": Object {
        "id": "44",
        "name": "Beteigeuze",
      },
    },
  },
  "loaded": false,
  "loading": false,
  "result": Array [
    "42",
    "45",
  ],
  "selected": null,
  "updatedAt": null,
}
`;

exports[`createCollectionAdapter collectionAdapter reducers updateEmbedded updates the embedded entity 1`] = `
Object {
  "dirty": true,
  "entities": Object {
    "collection": Object {
      "42": Object {
        "firstname": "Arthur",
        "hobbies": Array [
          "12",
          "13",
        ],
        "id": "42",
        "lastname": "Dent",
        "planet": "43",
      },
      "45": Object {
        "firstname": "Tricia",
        "hobbies": Array [
          "12",
        ],
        "id": "45",
        "lastname": "McMillan",
        "planet": "43",
      },
    },
    "hobbies": Object {
      "12": Object {
        "id": "12",
        "name": "drinking coffee",
      },
      "13": Object {
        "id": "13",
        "name": "Carrying towels",
      },
    },
    "planets": Object {
      "43": Object {
        "id": "43",
        "name": "Earth",
      },
    },
  },
  "loaded": false,
  "loading": false,
  "result": Array [
    "42",
    "45",
  ],
  "selected": null,
  "updatedAt": null,
}
`;

exports[`createCollectionAdapter collectionAdapter reducers upsert adds the entity to an empty state 1`] = `
Object {
  "dirty": true,
  "entities": Object {
    "collection": Object {
      "42": Object {
        "firstname": "Arthur",
        "hobbies": Array [
          "12",
          "13",
        ],
        "id": "42",
        "lastname": "Dent",
        "planet": "43",
      },
    },
    "hobbies": Object {
      "12": Object {
        "id": "12",
        "name": "Drinking tea",
      },
      "13": Object {
        "id": "13",
        "name": "Carrying towels",
      },
    },
    "planets": Object {
      "43": Object {
        "id": "43",
        "name": "Earth",
      },
    },
  },
  "loaded": false,
  "loading": false,
  "result": Array [
    "42",
  ],
  "selected": null,
  "updatedAt": null,
}
`;

exports[`createCollectionAdapter collectionAdapter reducers upsertMany adds the entities to an empty state 1`] = `
Object {
  "dirty": true,
  "entities": Object {
    "collection": Object {
      "42": Object {
        "firstname": "Arthur",
        "hobbies": Array [
          "12",
          "13",
        ],
        "id": "42",
        "lastname": "Dent",
        "planet": "43",
      },
      "45": Object {
        "firstname": "Tricia",
        "hobbies": Array [
          "12",
        ],
        "id": "45",
        "lastname": "McMillan",
        "planet": "43",
      },
    },
    "hobbies": Object {
      "12": Object {
        "id": "12",
        "name": "Drinking tea",
      },
      "13": Object {
        "id": "13",
        "name": "Carrying towels",
      },
    },
    "planets": Object {
      "43": Object {
        "id": "43",
        "name": "Earth",
      },
    },
  },
  "loaded": false,
  "loading": false,
  "result": Array [
    "42",
    "45",
  ],
  "selected": null,
  "updatedAt": null,
}
`;
